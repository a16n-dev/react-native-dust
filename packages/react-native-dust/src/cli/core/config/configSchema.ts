import { z } from 'zod';

const themeSchema = z
  .object({
    colors: z.record(z.string(), z.record(z.string(), z.string())),
    spacing: z.record(z.string(), z.number()),
    radius: z.record(z.string(), z.number()),
    shadow: z.record(z.string(), z.string()),
    text: z.record(
      z.string(),
      z.object({
        fontSize: z.number(),
        lineHeight: z.number().optional(),
        letterSpacing: z.number().optional(),
      })
    ),
  })
  .loose();

const extendedThemeSchema = z.object({
  extend: themeSchema.partial(),
});

export const configSchema = z
  .object({
    include: z.array(z.string()),
    theme: themeSchema,
    additionalThemes: z
      .record(z.string(), z.union([themeSchema, extendedThemeSchema]))
      .optional(),
    breakpoints: z.record(z.string(), z.number()).optional(),
    options: z
      .object({
        targetsWeb: z.boolean().optional(),
        mode: z.enum(['vanilla', 'unistyles']).default('vanilla'),
        tokenAllowList: z.array(z.string()).default([]),
        tokenBlockList: z.array(z.string()).default([]),
        generatePlatformHelpers: z.boolean().default(true),
      })
      .prefault({}),
  })
  .refine(
    (data) => {
      // If additionalThemes exists, mode must be "unistyles"
      if (data.additionalThemes && data.options.mode !== 'unistyles')
        return false;
      return true;
    },
    {
      message: "additionalThemes can only be present when mode is 'unistyles'",
      path: ['additionalThemes'],
    }
  );

export type ParsedConfig = z.infer<typeof configSchema>;
export type ParsedTheme = z.infer<typeof themeSchema>;
export type ParsedExtendedTheme = z.infer<typeof extendedThemeSchema>;

export type InputConfig = z.input<typeof configSchema>;
export type InputTheme = z.input<typeof themeSchema>;
export type InputExtendedTheme = z.input<typeof extendedThemeSchema>;
