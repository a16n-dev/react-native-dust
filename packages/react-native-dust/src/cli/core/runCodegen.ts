import { generateThemeFile } from '../templates/generateThemeFile.js';
import type { codegenOptions } from './codegenTypes.js';
import { GeneratedProject } from '../templates/generatedProject.js';
import { generateTokensFile } from '../templates/generateTokensFile.js';
import { generateUtilitiesFile } from '../templates/generateUtilitiesFile.js';
import { generateBarrelFile } from '../templates/generateBarrelFile.js';
import type { ParsedConfig } from './config/configSchema.js';
import { writeGeneratedLibFiles } from './fileSystemHelpers.js';
import { logger } from '../logger/logger.js';
import { c } from '../logger/format.js';

export async function constructCodegenProject(opts: codegenOptions) {
  const project = new GeneratedProject();

  // Add project
  generateThemeFile(project, opts);
  generateTokensFile(project, opts);
  if (opts.config.options.generatePlatformHelpers) {
    generateUtilitiesFile(project, opts);
  }
  generateBarrelFile(project, opts);

  return project;
}

export async function runCodegen(
  config: ParsedConfig,
  whitelist?: string[]
): Promise<void> {
  const project = await constructCodegenProject({ config, whitelist });

  const outputFiles = project.getGeneratedFiles();

  const errors = project.typeCheck();
  if (errors) {
    // TODO: If this error is hit, it indicates a bug in the codegen rather than user error
    // We should probably get the user to file a bug report, or maybe get this to fire off
    // a network request somewhere to log the error automatically
    logger.error(
      c.red('Error: Failed to generate dynamic code due to type errors')
    );
    logger.error(errors);
    logger.error(
      'The above error is likely due to a bug in the library rather than your code or configuration. Please report the issue here: https://github.com/a16n-dev/react-native-dust/issues'
    );
  }

  await writeGeneratedLibFiles(outputFiles);
}
