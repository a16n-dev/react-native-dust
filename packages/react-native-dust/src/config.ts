export type DustTheme = {
  colors: Record<string, Record<string, string>>;
  spacing: Record<string, number>;
  radius: Record<string, number>;
  shadow: Record<string, string>;
  text: Record<
    string,
    {
      fontSize: number;
      lineHeight?: number;
      letterSpacing?: number;
    }
  >;
  [key: string]: any;
  extend?: never;
};

export type Config = {
  /**
   * An array of glob patterns or directory paths to include in the code generation process.
   * These patterns determine which files will be scanned for token usage by `dust build`
   */
  include: string[];
  theme: DustTheme;
  breakpoints?: Record<string, number>;
  options?: {
    /**
     * If true, will generate a set of web-only tokens, such as web_fixed.
     */
    targetsWeb?: boolean;
    /**
     * Determines the mode for the generated styles. Defaults to 'vanilla'.
     * - 'vanilla': Single theme support, plain stylesheet output, no runtime dependencies
     * - 'unistyles': Multiple themes support, safe area margin/padding utility classes. Requires unistyles.
     */
    mode?: 'vanilla' | 'unistyles';
    /**
     * Specify tokens that should always be generated by `dust build`, even if they are not used in the codebase.
     * Can be specified as an array of token strings, or regular expressions as strings.
     */
    tokenAllowList?: string[];
    /**
     * Specify tokens that should never be generated by `dust build`. Use this if you want to exclude certain
     * default tokens.
     */
    tokenBlockList?: string[];
    /**
     * Generate a set of utility helper functions for platform-specific code such as `isIOS`, `android()`
     * Defaults to true
     */
    generatePlatformHelpers?: boolean;
  };
};
