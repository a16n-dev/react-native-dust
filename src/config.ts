export type DustTheme = {
  colors: Record<string, Record<string, string>>;
  spacing: Record<string, number>;
  radius: Record<string, number>;
  shadow: Record<string, string>;
  text: Record<
    string,
    {
      fontSize: number;
      lineHeight?: number;
      letterSpacing?: number;
      fontWeight?: string | number;
    }
  >;
  [key: string]: any;
};

export type ExtendedDustTheme = { extend: Partial<DustTheme> };

export type Config = {
  include: string[];
  theme: DustTheme;
  additionalThemes?: Record<string, DustTheme | ExtendedDustTheme>;
  breakpoints?: Record<string, number>;
  options?: {
    /**
     * If true, will generate a set of web-only tokens, such as web_fixed.
     */
    targetsWeb?: boolean;
    /**
     * Determines the mode for the generated styles. Defaults to 'vanilla'.
     * - 'vanilla': Single theme support, plain stylesheet output, no runtime dependencies
     * - 'unistyles': Multiple themes support, safe area margin/padding utility classes. Requires unistyles.
     */
    mode?: "vanilla" | "unistyles";
  };
};
