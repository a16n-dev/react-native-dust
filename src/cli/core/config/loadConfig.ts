import { existsSync } from 'fs';
import { resolve } from 'path';
import { createJiti } from 'jiti';
import { z } from 'zod';
import { configSchema, ParsedConfig } from './configSchema';

/**
 * These are the locations where we expect to find a config file by default, relative to the root of the project.
 */
const DEFAULT_CONFIG_FILES = [
  'dust.config.ts',
  'dust.config.js',
  'dust.config.json',
];

/**
 * This loads the config for the dust CLI. If no config path is provided, this function will attempt to read a config
 * file at the default locations - dust.config.(json|js|ts).
 *
 * It will also validate that the config matches the expected type,
 */
export function loadConfig(configPath?: string): ParsedConfig {
  const resolvedPath = findConfigFile(configPath);

  if (!resolvedPath) {
    if (!configPath) {
      console.error('Error: No config file found. Looked for:');
      DEFAULT_CONFIG_FILES.forEach((file) => console.error(`  - ${file}`));
    } else {
      console.error(`Error: Config file not found at ${resolve(configPath)}`);
    }
    process.exit(1);
  }

  try {
    // jiti can load both .js and .ts files seamlessly
    const jiti = createJiti(__filename);
    const config = jiti(resolvedPath);

    const rawConfigJson = config.default || config;

    // Validate the config
    const parseResult = configSchema.safeParse(rawConfigJson);

    if (parseResult.error) {
      console.error('Invalid config file: ');
      console.error(z.prettifyError(parseResult.error));
      process.exit(1);
    }

    return parseResult.data;
  } catch (error) {
    console.error(`Error loading config file: ${error}`);
    process.exit(1);
  }
}

function findConfigFile(configPath?: string): string | null {
  // If a specific path was provided, use it as-is
  if (configPath) {
    return resolve(configPath);
  }

  // Check for default config files in order of preference
  for (const config of DEFAULT_CONFIG_FILES) {
    const resolvedPath = resolve(config);
    if (existsSync(resolvedPath)) {
      return resolvedPath;
    }
  }

  return null;
}
