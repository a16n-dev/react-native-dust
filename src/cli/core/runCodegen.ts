import { writeGeneratedLibFiles } from './uiWriter';
import { generateVanillaThemeFile } from '../templates/generateVanillaThemeFile';
import { generateVanillaTokensFile } from '../templates/generateVanillaTokensFile';
import { GeneratedProject } from '../templates/getGeneratedSource';
import { generateBarrelFile } from '../templates/generateBarrelFile';
import { codegenOptions } from './codegenTypes';
import { ParsedConfig } from './loadConfig';

export function constructCodegenProject(opts: codegenOptions) {
  const project = new GeneratedProject();

  // Add project
  generateVanillaThemeFile(project, opts);
  generateVanillaTokensFile(project, opts);
  generateBarrelFile(project);

  return project;
}

export async function runCodegen(
  config: ParsedConfig,
  whitelist?: string[]
): Promise<void> {
  const project = constructCodegenProject({ config, whitelist });

  const outputFiles = project.getGeneratedFiles();

  const errors = project.typeCheck();
  if (errors) {
    // TODO: If this error is hit, it indicates a bug in the codegen rather than user error
    // We should probably get the user to file a bug report, or maybe get this to fire off
    // a network request somewhere to log the error automatically
    console.error(errors);
    process.exit(1);
  }

  await writeGeneratedLibFiles(outputFiles);
}
