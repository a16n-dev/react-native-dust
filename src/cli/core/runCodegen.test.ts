import { describe, expect, test } from 'vitest';
import { constructCodegenProject } from './runCodegen';
import { ParsedConfig } from './config/configSchema';

const TEST_VANILLA_CONFIG: ParsedConfig = {
  include: ['src'],
  options: {
    mode: 'vanilla',
  },
  theme: {
    colors: {},
    radius: {},
    shadow: {},
    spacing: {},
    text: {},
  },
};

describe('constructCodegenProject', () => {
  test('should construct the expected output for the first sample configuration', async () => {
    const project = await constructCodegenProject({
      config: TEST_VANILLA_CONFIG,
    });

    expect(project.getSourceFiles()).toMatchSnapshot();
    expect(project.getGeneratedFiles()).toMatchSnapshot();
  });
});
