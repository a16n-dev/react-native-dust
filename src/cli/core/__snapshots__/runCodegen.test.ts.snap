// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`constructCodegenProject > should construct the expected output for the first sample configuration 1`] = `
[
  {
    "content": "export * from "./theme";
export * from "./tokens";
",
    "name": "index.ts",
  },
  {
    "content": "interface AppTheme {
}

export const theme = {
      "colors": {},
      "radius": {},
      "shadow": {},
      "spacing": {},
      "text": {}
    };
",
    "name": "theme.ts",
  },
  {
    "content": "import { StyleSheet } from "react-native";
import { theme } from "./theme";
",
    "name": "tokens.ts",
  },
]
`;

exports[`constructCodegenProject > should construct the expected output for the first sample configuration 2`] = `
[
  {
    "content": "export const theme = {
    "colors": {},
    "radius": {},
    "shadow": {},
    "spacing": {},
    "text": {}
};
",
    "name": "theme.js",
  },
  {
    "content": "export declare const theme: {
    colors: {};
    radius: {};
    shadow: {};
    spacing: {};
    text: {};
};
",
    "name": "theme.d.ts",
  },
  {
    "content": "export {};
",
    "name": "tokens.js",
  },
  {
    "content": "export {};
",
    "name": "tokens.d.ts",
  },
  {
    "content": "export * from "./theme";
export * from "./tokens";
",
    "name": "index.js",
  },
  {
    "content": "export * from "./theme";
export * from "./tokens";
",
    "name": "index.d.ts",
  },
]
`;
